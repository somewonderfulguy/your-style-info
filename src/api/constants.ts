export const pagePathPair = new Map([
  ['/en', '/json/home-en.json'],
  ['/ru', '/json/home-ru.json'],
  ['/en/guides-topics', '/json/guides-en.json'],
  ['/ru/guides-topics', '/json/guides-ru.json'],
  ['/en/guides-topics/colors-how-to', '/json/colors-how-to-en.json'],
  ['/ru/guides-topics/colors-how-to', '/json/colors-how-to-ru.json'],
  ['/en/guides-topics/how-to-travel', '/json/how-to-travel-en.json'],
  ['/ru/guides-topics/how-to-travel', '/json/how-to-travel-ru.json'],
  ['/en/guides-topics/black-white-ties', '/json/black-white-ties-en.json'],
  ['/ru/guides-topics/black-white-ties', '/json/black-white-ties-ru.json'],
  ['/en/clothes', '/json/clothes-en.json'],
  ['/ru/clothes', '/json/clothes-ru.json'],
  ['/en/clothes/suits-jackets-vests', '/json/suits-jackets-vests-en.json'],
  ['/ru/clothes/suits-jackets-vests', '/json/suits-jackets-vests-ru.json'],
  ['/en/clothes/shirts', '/json/shirts-en.json'],
  ['/ru/clothes/shirts', '/json/shirts-ru.json'],
  ['/en/clothes/knitwear', '/json/knitwear-en.json'],
  ['/ru/clothes/knitwear', '/json/knitwear-ru.json'],
  ['/en/clothes/shoes', '/json/shoes-en.json'],
  ['/ru/clothes/shoes', '/json/shoes-ru.json'],
  ['/en/clothes/outerwear', '/json/outerwear-en.json'],
  ['/ru/clothes/outerwear', '/json/outerwear-ru.json'],
  ['/en/clothes/accessories', '/json/accessories-en.json'],
  ['/ru/clothes/accessories', '/json/accessories-ru.json'],
  ['/en/clothes/materials-and-fabrics', '/json/materials-and-fabrics-en.json'],
  ['/ru/clothes/materials-and-fabrics', '/json/materials-and-fabrics-ru.json'],
  ['/en/clothes/patterns', '/json/patterns-en.json'],
  ['/ru/clothes/patterns', '/json/patterns-ru.json'],
  ['/en/clothes/maintenance', '/json/maintenance-en.json'],
  ['/ru/clothes/maintenance', '/json/maintenance-ru.json'],
  ['/en/grooming', '/json/grooming-en.json'],
  ['/ru/grooming', '/json/grooming-ru.json'],
  ['/en/grooming/face-skin-care', '/json/face-skin-care-en.json'],
  ['/ru/grooming/face-skin-care', '/json/face-skin-care-ru.json'],
  ['/en/grooming/hands-care', '/json/hands-care-en.json'],
  ['/ru/grooming/hands-care', '/json/hands-care-ru.json'],
  ['/en/brands-stores', '/json/brands-stores-en.json'],
  ['/ru/brands-stores', '/json/brands-stores-ru.json'],
  ['/en/glossary', '/json/glossary-en.json'],
  ['/ru/glossary', '/json/glossary-ru.json']
])
