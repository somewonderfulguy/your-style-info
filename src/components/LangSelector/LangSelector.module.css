.langSelector {
  position: relative;
  display: inline-block;
}

.langSelectorDark {
  composes: langSelector;
}

.langSelector button {
  font-size: 15px;
  color: inherit;
  cursor: pointer;
}

.langSelectorDark button {
  font-size: 15px;
  color: #919191;
  cursor: pointer;
}

.gray button {
  color: #696969;
}

.langSelectorDark.gray button {
  color: #919191;
}
.langSelectorDark.gray > button {
  color: #696969;
}

.langSelectorInner {
  display: flex;
  align-items: center;

  opacity: 0.7;
  transition: opacity 200ms;
}

.langSelectorInner:hover {
  opacity: 1;
}

.langSelectorInner:disabled {
  opacity: 0.4;
}

.icon,
.langSelectorDark.gray .icon {
  margin-right: 5px;
  fill: #696969;
}

.langSelectorDark .icon {
  fill: #919191;
}

.langSelector .triangle {
  margin-left: 5px;
  width: 0;
  height: 0;
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid rgb(105, 105, 105);
  border-bottom: 0;
}

.triangleReverse {
  composes: triangle;
  transform: rotate(180deg);
}

.langSelector .dropDown {
  position: absolute;
  right: 0;
  z-index: 1;

  list-style: none;
  margin: 2px 0;
  padding: 8px 0;

  border: 1px solid rgba(var(--primaryBorder), 0.6);
  border-radius: 4px;
  background-color: rgb(var(--primaryBackground));
}

.dropDownAbove {
  composes: dropDown;
  top: 0;
  transform: translateY(calc(-100% - 8px));
}

.dropDownBelow {
  composes: dropDown;
  bottom: 0;
  transform: translateY(calc(100% + 4px));
}

.dropDown button {
  width: 100%;
  padding: 4px 24px;
  text-align: left;
  transition: background-color 0.2s;
}

.dropDown button:hover {
  background-color: rgba(var(--primaryHoverHighlight), 0.7);
}

.dropDown button:disabled {
  cursor: inherit;
  color: rgb(var(--primaryDisabled));
}

.dropDown button:disabled:hover {
  background-color: transparent;
}

/* TODO FIX ? */
.dropDown button:active {
  color: #fff;
  background-color: #007bff;
}
