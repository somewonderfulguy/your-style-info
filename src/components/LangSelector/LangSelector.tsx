import { useRef, useState } from 'react'

import { useThemeState } from '~contexts/themeContext'
import { useLocalization } from '~contexts/localizationContext'
import { useOutsideClick } from '~shared/hooks'
import { LanguageIcon } from '~assets/images'
import { LANGUAGES, LOCALES } from '~constants/index'

import styles from './LangSelector.module.css'

type Props = {
  showAbove?: boolean
  gray?: boolean
}

const LangSelector = ({ showAbove = false, gray = false }: Props) => {
  const [
    { locale: currentLocale },
    setLocaleState,
    { isLoading: isLanguageLoading, data: localeData }
  ] = useLocalization()

  const isDarkTheme = useThemeState()
  const langSelectorRef = useRef(null)
  const menuRef = useRef(null)
  const [isOpen, setOpen] = useState(false)

  useOutsideClick(langSelectorRef, () => setOpen(false))

  const langSelectorClass = isDarkTheme
    ? styles.langSelectorDark
    : styles.langSelector
  const grayClass = gray ? styles.gray : ''

  const triangleClass = showAbove
    ? isOpen
      ? styles.triangle
      : styles.triangleReverse
    : isOpen
      ? styles.triangleReverse
      : styles.triangle

  return (
    <div ref={langSelectorRef} className={`${langSelectorClass} ${grayClass}`}>
      <button
        onClick={() => setOpen((isOpen) => !isOpen)}
        className={styles.langSelectorInner}
        type="button"
        disabled={isLanguageLoading}
        aria-label={localeData?.switchLanguage.switchLang ?? 'Switch language'}
        aria-expanded={isOpen}
        aria-haspopup
        data-testid="langSelector"
      >
        <LanguageIcon width={20} height={20} className={styles.icon} />
        <span>{LANGUAGES.get(currentLocale)}</span>
        <div className={triangleClass} />
      </button>

      {isOpen && (
        <ul
          ref={menuRef}
          className={showAbove ? styles.dropDownAbove : styles.dropDownBelow}
          role="menu"
        >
          {LOCALES.map((locale) => (
            <li role="menuitem" key={locale}>
              <button
                type="button"
                disabled={locale === currentLocale}
                onClick={() => {
                  setLocaleState(locale as 'en' | 'ua')
                  setOpen(false)
                }}
              >
                {LANGUAGES.get(locale)}
              </button>
            </li>
          ))}
        </ul>
      )}
    </div>
  )
}

export default LangSelector
