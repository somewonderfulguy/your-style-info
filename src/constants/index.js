export * from './common'
export * from './componentTypes'
export * from './languages'
export * from './primeRoutes'