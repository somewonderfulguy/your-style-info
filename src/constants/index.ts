export * from './common'
export * from './languages'
export * from './rootMenuThumbs'
export * from './primeRoutes'