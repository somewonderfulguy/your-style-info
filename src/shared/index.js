export * from './componentRenderer'
export * from './getComponent'
export * from './tests'