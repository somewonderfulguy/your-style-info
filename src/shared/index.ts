export * from './hooks'
export * from './tests'
export * from './utils'

export * from './types'