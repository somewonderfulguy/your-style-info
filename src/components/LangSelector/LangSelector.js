import React, {useRef, useState} from 'react'
import {bool} from 'prop-types'

import {useLocalisation, useTheme} from 'contexts'
import {useOutsideClick} from 'shared/hooks'
import {LanguageIcon} from 'assets/images'
import {LANGUAGES, LOCALES} from 'constants/index'
import styles from './LangSelector.module.css'

const propTypes = {
  showAbove: bool,
  gray: bool
}

const defaultProps = {
  showAbove: false,
  gray: false
}

const LangSelector = ({showAbove, gray}) => {
  const {locale: currentLocale, setLocale} = useLocalisation()
  const {isDarkTheme} = useTheme()
  const langSelectorRef = useRef(null)
  const menuRef = useRef(null)
  const [isOpen, setOpen] = useState(false)

  useOutsideClick(langSelectorRef, () => setOpen(false))

  const langSelectorClass = isDarkTheme ? styles.langSelectorDark : styles.langSelector
  const grayClass = gray ? styles.gray : ''

  const triangleClass = showAbove
    ? (isOpen ? styles.triangle : styles.triangleReverse)
    : (isOpen ? styles.triangleReverse : styles.triangle)

  return (
    <div
      ref={langSelectorRef}
      className={`${langSelectorClass} ${grayClass}`}
    >
      <button
        onClick={() => setOpen(isOpen => !isOpen)}
        className={styles.langSelectorInner}
        type="button"
        // TODO switch label depending on language
        aria-label="Switch language"
        aria-expanded={isOpen}
        aria-haspopup
      >
        <LanguageIcon width={20} height={20} className={styles.icon} />
        <span>{LANGUAGES.get(currentLocale)}</span>
        <div className={triangleClass} />
      </button>

      {isOpen && (
        <ul
          ref={menuRef}
          className={showAbove ? styles.dropDownAbove : styles.dropDownBelow}
          role="menu"
        >
          {LOCALES.map(locale => (
            <li role="menuitem" key={locale}>
              <button
                type="button"
                disabled={locale === currentLocale}
                onClick={() => {
                  setLocale(locale)
                  setOpen(false)
                }}
              >
                {LANGUAGES.get(locale)}
              </button>
            </li>
          ))}
        </ul>
      )}
    </div>
  )
}

LangSelector.propTypes = propTypes
LangSelector.defaultProps = defaultProps

export default LangSelector