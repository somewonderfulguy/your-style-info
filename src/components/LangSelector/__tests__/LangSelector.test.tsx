import React from 'react'

import {act, fireEvent, render, renderWholeApp, userEvent, screen, waitFor} from 'shared/tests'
import LangSelector from '..'
import {FULL_LOCALES} from 'constants/index'

expect.addSnapshotSerializer({
  test(val) {return true},
  print(val) {return (val as string).replace(/\n\+\s{0,}<ul(\n.*){1,}/gm, '')}
})

const [english, russian] = FULL_LOCALES
const englishRegExp = new RegExp(english, 'i')
const russianRegExp = new RegExp(russian, 'i')

const setup = (showAbove?: boolean, gray?: boolean, path = '/') => {
  const utils = render(
    <LangSelector showAbove={showAbove} gray={gray} />,
    {route: path}
  )

  const getSwitchLanguageBtn = () => screen.getByTestId('langSelector')

  return {
    ...utils,
    getSwitchLanguageBtn
  }
}

test('snapshot diff: aria state and css-class of triangle when open/hidden menu (+ drop-up/down diff)', async () => {
  const diffSnapshotOptions = {
    contextLines: 1,
    aAnnotation: 'collapsed menu',
    bAnnotation: 'expanded menu'
  }

  const {getSwitchLanguageBtn, asFragment, rerender} = setup()
  const beforeClick = asFragment()
  const switchLanguageBtn = getSwitchLanguageBtn()

  const btn = screen.getByRole('button', {
    name: /switch language/i
  })

  await waitFor(() => expect(btn).not.toBeDisabled())

  userEvent.click(btn)
  expect(beforeClick).toMatchDiffSnapshot(asFragment(), diffSnapshotOptions)

  userEvent.click(switchLanguageBtn)
  rerender(<LangSelector showAbove />)

  const beforeClickShowAbove = asFragment()
  userEvent.click(switchLanguageBtn)
  expect(beforeClickShowAbove).toMatchDiffSnapshot(asFragment(), diffSnapshotOptions)
})

test('should show/hide menu when clicking on language selector', async () => {
  const url = '/en/outerwear/trench-coat'
  const {getSwitchLanguageBtn} = setup(false, false, url)
  const langSelectorBtn = getSwitchLanguageBtn()

  expect(screen.queryByRole('menu')).not.toBeInTheDocument()

  // open
  await waitFor(() => expect(langSelectorBtn).not.toBeDisabled())
  userEvent.click(langSelectorBtn)
  expect(screen.getByRole('menu')).toBeInTheDocument()

  // select language - hides menu (also check that button text changes depending on current lang, and current lang in menu must be disabled)
  const getEnglishMenuItem = () => screen.getByRole('menuitem', {name: englishRegExp}).querySelector('button')
  const getRussianMenuItem = () => screen.getByRole('menuitem', {name: russianRegExp}).querySelector('button')

  expect(langSelectorBtn).toHaveTextContent(new RegExp(english, 'i'))
  expect(getEnglishMenuItem()).toBeDisabled()
  expect(getRussianMenuItem()).not.toBeDisabled()
  userEvent.click(screen.getByRole('menu').querySelector('button:not(:disabled)') as HTMLElement)
  expect(screen.queryByRole('menu')).not.toBeInTheDocument()
  await waitFor(() => expect(langSelectorBtn).toHaveTextContent(russianRegExp))

  // open
  await waitFor(() => expect(langSelectorBtn).not.toBeDisabled())
  userEvent.click(langSelectorBtn)
  expect(getEnglishMenuItem()).not.toBeDisabled()
  expect(getRussianMenuItem()).toBeDisabled()

  // close
  userEvent.click(langSelectorBtn)
  expect(screen.queryByRole('menu')).not.toBeInTheDocument()

  // close by clicking outside
  userEvent.click(langSelectorBtn)
  act(() => userEvent.click(document.body))
  expect(screen.queryByRole('menu')).not.toBeInTheDocument()
})

// eslint-disable-next-line jest/expect-expect
test('header and footer lang selectors, light/dark themes, desktop/mobile', async () => {
  renderWholeApp()

  const diffSnapshotOptions = {
    contextLines: 1,
    aAnnotation: 'light theme',
    bAnnotation: 'dark theme'
  }

  const [blockHeaderLangSelector, blockFooterLangSelector] = screen.getAllByText(russianRegExp)
  const headerLangSelector = blockHeaderLangSelector.closest('button')
  const footerLangSelector = blockFooterLangSelector.closest('button')
  const headerLangSelectorWrapper = headerLangSelector?.parentElement
  const footerLangSelectorWrapper = footerLangSelector?.parentElement
  const getMenu = () => screen.getByRole('menu')

  // theme switcher from footer as it always visible
  const [, themeSwitcher] = screen.getAllByLabelText(/switch theme/i)

  userEvent.click(headerLangSelector as HTMLButtonElement)
  expect(getMenu().className).toMatchSnapshot('desktop header drop down')

  userEvent.click(footerLangSelector as HTMLButtonElement)
  expect(getMenu().className).toMatchSnapshot('desktop footer drop up')

  // desktop header theme
  const desktopHeaderLight = headerLangSelectorWrapper?.className
  userEvent.click(themeSwitcher)
  const desktopHeaderDark = headerLangSelectorWrapper?.className
  expect(desktopHeaderLight).toMatchDiffSnapshot(desktopHeaderDark, diffSnapshotOptions)

  // desktop footer theme
  const desktopFooterDark = footerLangSelectorWrapper?.className
  userEvent.click(themeSwitcher)
  const desktopFooterLight = footerLangSelectorWrapper?.className
  expect(desktopFooterLight).toMatchDiffSnapshot(desktopFooterDark, diffSnapshotOptions)

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  window.innerWidth = 320
  fireEvent(window, new Event('resize'))

  // note: mobile header is in another component - Options
  expect(headerLangSelector).not.toBeInTheDocument()

  userEvent.click(footerLangSelector as HTMLButtonElement)
  expect(getMenu().className).toMatchSnapshot('mobile footer drop up')

  // mobile footer theme
  const mobileFooterLight = footerLangSelectorWrapper?.className
  userEvent.click(themeSwitcher)
  const mobileFooterDark = footerLangSelectorWrapper?.className

  expect(mobileFooterLight).toMatchDiffSnapshot(mobileFooterDark, diffSnapshotOptions)
})