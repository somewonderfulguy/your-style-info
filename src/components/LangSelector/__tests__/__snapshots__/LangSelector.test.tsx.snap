// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header and footer lang selectors, light/dark themes, desktop/mobile 1`] = `
Snapshot Diff:
- light theme
+ dark theme

- langSelector 
+ langSelectorDark 
`;

exports[`header and footer lang selectors, light/dark themes, desktop/mobile 2`] = `
Snapshot Diff:
- light theme
+ dark theme

- langSelector gray
+ langSelectorDark gray
`;

exports[`header and footer lang selectors, light/dark themes, desktop/mobile 3`] = `
Snapshot Diff:
- light theme
+ dark theme

- langSelector gray
+ langSelectorDark gray
`;

exports[`header and footer lang selectors, light/dark themes, desktop/mobile: desktop footer drop up 1`] = `dropDownAbove`;

exports[`header and footer lang selectors, light/dark themes, desktop/mobile: desktop header drop down 1`] = `dropDownBelow`;

exports[`header and footer lang selectors, light/dark themes, desktop/mobile: mobile footer drop up 1`] = `dropDownAbove`;

exports[`snapshot diff: aria state and css-class of triangle when open/hidden menu (+ drop-up/down diff) 1`] = `
Snapshot Diff:
- collapsed menu
+ expanded menu

@@ -5,3 +5,3 @@
      <button
-       aria-expanded="false"
+       aria-expanded="true"
        aria-haspopup="true"
@@ -10,3 +10,2 @@
        data-testid="langSelector"
-       disabled=""
        type="button"
@@ -27,5 +26,29 @@
        <div
-         class="triangle"
+         class="triangleReverse"
        />
      </button>
`;

exports[`snapshot diff: aria state and css-class of triangle when open/hidden menu (+ drop-up/down diff) 2`] = `
Snapshot Diff:
- collapsed menu
+ expanded menu

@@ -5,3 +5,3 @@
      <button
-       aria-expanded="false"
+       aria-expanded="true"
        aria-haspopup="true"
@@ -26,5 +26,29 @@
        <div
-         class="triangleReverse"
+         class="triangle"
        />
+     </button>
`;
